### DISCUSSION TITLE
Response time of the use case of Paho Java

### DISCUSSION THREAD

**Discussion by gtahubb** [emqx/emqx#4729] [Q&A]
**the dependency is org.eclipse.paho.client.mqttv3. If it is 1.2.0 and the previous version, the response time is around 50ms, and if it is a later version, it is 350ms.**:

**Comment by k32** [comment]
Hello,
I assume the latency changed after you updated paho plugin in your client application? https://mvnrepository.com/artifact/org.eclipse.paho/org.eclipse.paho.client.mqttv3

Did you change anything in the EMQ X installation?

**Reply by gtahubb** [reply]
**Default** configuration
**mqtt_version,docker** images: emqx/emqx-ee  4.2.2

**Java:**
<parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.2.7.RELEASE</version>
    </parent>

    <dependencies>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-web</artifactId>
        </dependency>

        <dependency>
            <groupId>org.eclipse.paho</groupId>
            <artifactId>org.eclipse.paho.client.mqttv3</artifactId>
            <version>1.2.1</version>
        </dependency>



    </dependencies>



@RestController
@RequestMapping("/mqtt")
public class SendController {

    @RequestMapping("/message")
    public Map sendMessage(){
        String subTopic = "testtopic/#";
        String pubTopic = "testtopic/1";
        String content = "Hello World";
        int qos = 2;
        String broker = "tcp://192.168.8.128:1883";
        String clientId = UUID.randomUUID().toString();
        MemoryPersistence persistence = new MemoryPersistence();

        try {
            MqttClient client = new MqttClient(broker, clientId, persistence);

            // MQTT 连接选项
            MqttConnectOptions connOpts = new MqttConnectOptions();
            connOpts.setUserName("admin");
            connOpts.setPassword("public".toCharArray());
            // 保留会话
            connOpts.setCleanSession(true);

            // 设置回调
            client.setCallback(new OnMessageCallback());

            // 建立连接
            System.out.println("Connecting to broker: " + broker);
            client.connect(connOpts);

            System.out.println("Connected");
            System.out.println("Publishing message: " + content);

            // 订阅
            //client.subscribe(subTopic);

            // 消息发布所需参数
            MqttMessage message = new MqttMessage(content.getBytes());
            message.setQos(qos);
            client.publish(pubTopic, message);
            System.out.println("Message published");

            client.disconnect();
            System.out.println("Disconnected");
            client.close();
            //System.exit(0);
        } catch (MqttException me) {
            System.out.println("reason " + me.getReasonCode());
            System.out.println("msg " + me.getMessage());
            System.out.println("loc " + me.getLocalizedMessage());
            System.out.println("cause " + me.getCause());
            System.out.println("excep " + me);
            me.printStackTrace();
        }
        return new HashMap();
    }

}


@RestController
@RequestMapping("/send")
public class SendController {

    @RequestMapping("/message")
    public Map sendMessage(){
        String subTopic = "testtopic/#";
        String pubTopic = "testtopic/1";
        String content = "Hello World";
        int qos = 2;
        String broker = "tcp://192.168.8.128:1883";
        String clientId = UUID.randomUUID().toString();
        MemoryPersistence persistence = new MemoryPersistence();

        try {
            MqttClient client = new MqttClient(broker, clientId, persistence);

            // MQTT 连接选项
            MqttConnectOptions connOpts = new MqttConnectOptions();
            connOpts.setUserName("admin");
            connOpts.setPassword("public".toCharArray());
            // 保留会话
            connOpts.setCleanSession(true);

            // 设置回调
            client.setCallback(new OnMessageCallback());

            // 建立连接
            System.out.println("Connecting to broker: " + broker);
            client.connect(connOpts);

            System.out.println("Connected");
            System.out.println("Publishing message: " + content);

            // 订阅
            //client.subscribe(subTopic);

            // 消息发布所需参数
            MqttMessage message = new MqttMessage(content.getBytes());
            message.setQos(qos);
            client.publish(pubTopic, message);
            System.out.println("Message published");

            client.disconnect();
            System.out.println("Disconnected");
            client.close();
            //System.exit(0);
        } catch (MqttException me) {
            System.out.println("reason " + me.getReasonCode());
            System.out.println("msg " + me.getMessage());
            System.out.println("loc " + me.getLocalizedMessage());
            System.out.println("cause " + me.getCause());
            System.out.println("excep " + me);
            me.printStackTrace();
        }
        return new HashMap();
    }

}

@SpringBootApplication
public class MqttApplication {
    public static void main(String[] args) {
        SpringApplication.run(MqttApplication.class,args);
    }
}

**Comment by gtahubb** [comment]
> Hello,
> I assume the latency changed after you updated paho plugin in your client application? https://mvnrepository.com/artifact/org.eclipse.paho/org.eclipse.paho.client.mqttv3
> 
> Did you change anything in the EMQ X installation?
Download it and use it directly

**Reply by gtahubb** [reply]
org.eclipse.paho.client.mqttv3. 1.2.0/1.2.1 Version response time
Default configuration
mqtt_version,docker images: emqx/emqx-ee 4.2.2

Java:

org.springframework.boot
spring-boot-starter-parent
2.2.7.RELEASE
<dependencies>

    <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-web</artifactId>
    </dependency>

    <dependency>
        <groupId>org.eclipse.paho</groupId>
        <artifactId>org.eclipse.paho.client.mqttv3</artifactId>
        <version>1.2.1</version>
    </dependency>
</dependencies>
@RestController
@RequestMapping("/mqtt")
public class SendController {

@RequestMapping("/message")
public Map sendMessage(){
    String subTopic = "testtopic/#";
    String pubTopic = "testtopic/1";
    String content = "Hello World";
    int qos = 2;
    String broker = "tcp://192.168.8.128:1883";
    String clientId = UUID.randomUUID().toString();
    MemoryPersistence persistence = new MemoryPersistence();

    try {
        MqttClient client = new MqttClient(broker, clientId, persistence);

        // MQTT 连接选项
        MqttConnectOptions connOpts = new MqttConnectOptions();
        connOpts.setUserName("admin");
        connOpts.setPassword("public".toCharArray());
        // 保留会话
        connOpts.setCleanSession(true);

        // 设置回调
        client.setCallback(new OnMessageCallback());

        // 建立连接
        System.out.println("Connecting to broker: " + broker);
        client.connect(connOpts);

        System.out.println("Connected");
        System.out.println("Publishing message: " + content);

        // 订阅
        //client.subscribe(subTopic);

        // 消息发布所需参数
        MqttMessage message = new MqttMessage(content.getBytes());
        message.setQos(qos);
        client.publish(pubTopic, message);
        System.out.println("Message published");

        client.disconnect();
        System.out.println("Disconnected");
        client.close();
        //System.exit(0);
    } catch (MqttException me) {
        System.out.println("reason " + me.getReasonCode());
        System.out.println("msg " + me.getMessage());
        System.out.println("loc " + me.getLocalizedMessage());
        System.out.println("cause " + me.getCause());
        System.out.println("excep " + me);
        me.printStackTrace();
    }
    return new HashMap();
}
}

@RestController
@RequestMapping("/send")
public class SendController {

@RequestMapping("/message")
public Map sendMessage(){
    String subTopic = "testtopic/#";
    String pubTopic = "testtopic/1";
    String content = "Hello World";
    int qos = 2;
    String broker = "tcp://192.168.8.128:1883";
    String clientId = UUID.randomUUID().toString();
    MemoryPersistence persistence = new MemoryPersistence();

    try {
        MqttClient client = new MqttClient(broker, clientId, persistence);

        // MQTT 连接选项
        MqttConnectOptions connOpts = new MqttConnectOptions();
        connOpts.setUserName("admin");
        connOpts.setPassword("public".toCharArray());
        // 保留会话
        connOpts.setCleanSession(true);

        // 设置回调
        client.setCallback(new OnMessageCallback());

        // 建立连接
        System.out.println("Connecting to broker: " + broker);
        client.connect(connOpts);

        System.out.println("Connected");
        System.out.println("Publishing message: " + content);

        // 订阅
        //client.subscribe(subTopic);

        // 消息发布所需参数
        MqttMessage message = new MqttMessage(content.getBytes());
        message.setQos(qos);
        client.publish(pubTopic, message);
        System.out.println("Message published");

        client.disconnect();
        System.out.println("Disconnected");
        client.close();
        //System.exit(0);
    } catch (MqttException me) {
        System.out.println("reason " + me.getReasonCode());
        System.out.println("msg " + me.getMessage());
        System.out.println("loc " + me.getLocalizedMessage());
        System.out.println("cause " + me.getCause());
        System.out.println("excep " + me);
        me.printStackTrace();
    }
    return new HashMap();
}
}

@SpringBootApplication
public class MqttApplication {
public static void main(String[] args) {
SpringApplication.run(MqttApplication.class,args);
}
}

**Reply by gtahubb** [reply]
https://docs.emqx.cn/broker/v4.3/development/java.html#paho-java-%E4%BD%BF%E7%94%A8%E7%A4%BA%E4%BE%8B

**Comment by k32** [comment]
Could you please clarify the problem? Do I understand correctly that you updated the dependency of your java application, and it changed the latency.
Did you change version of EMQ X, or some configuration of the broker?

**Comment by qzhuyan** [comment]
If possible, could you send us the tcpdump file so that we can check if there are any differences in the MQTT message?
